# Merge All Data

Merge air quality, AEMET climate and ERA5-Land data for each station.

In [1]:
# Load packages
suppressMessages(library(tidyverse))
suppressMessages(library(lubridate))

In [2]:
# Working directory
setwd("~/Repositories/AirQualityCOVID")

# Source some scripts 
source("src/functions.R")

### Main Variables

In [3]:
start_dt <- ymd_hms("2013-01-01 00:00:00")
end_dt <- ymd_hms("2020-12-31 00:00:00")

## Stations Information

In [4]:
#-----------------------------
#    Air Quality Stations
#-----------------------------

sites.AQ <- read.csv("data/Curation/checked_AQ.csv",
                    stringsAsFactor=T)

#-----------------------------
#       AEMET Stations
#-----------------------------

sites.AEMET <- read.csv("data/Curation/checked_AEMET.csv",
                        stringsAsFactor=F, colClasses=c("indicativo"="character"))

#-----------------------------
#       WorldMet Stations
#-----------------------------

sites.WorldMet <- read.csv("data/Curation/checked_WorldMet.csv",
                     stringsAsFactor=T)

In [5]:
head(sites.AEMET)

latitud  provincia altitud indicativo nombre             indsinop longitud 
1 43.29806 BIZKAIA    42     1082       BILBAO AEROPUERTO  8025     -2.906389
2 43.29806 BIZKAIA    42     1082       BILBAO AEROPUERTO  8025     -2.906389
3 40.41194 MADRID    667     3195       MADRID, RETIRO     8222     -3.678056
4 40.41194 MADRID    667     3195       MADRID, RETIRO     8222     -3.678056
5 37.41667 SEVILLA    34     5783       SEVILLA AEROPUERTO 8391     -5.879167
6 37.41667 SEVILLA    34     5783       SEVILLA AEROPUERTO 8391     -5.879167
  dist      siteAQ 
1  5.396849 es0041a
2  5.756578 es0110a
3  1.137154 es0118a
4  4.417472 es0120a
5  7.968651 es0817a
6 11.220392 es0890a

## n-Table 

Table with all the relations between Air Quality stations with AEMET and WorldMet stations code

| siteAQ  | AEMET |   WorldMet   |
|---------|-------|--------------|
| es1580a | 1111  | 080210-99999 |
| es0118a | 3195  | 082230-99999 |
| es1438a | 0201D | 081810-99999 |
| es1340a | 4642E | 083830-99999 |

In [6]:
nn.stations <- data.frame()

for (st in levels(sites.AQ$site)) {
    nn.stations <- rbind(nn.stations,
                         data.frame(siteAQ = as.factor(st),
                                    AEMET = as.factor(sites.AEMET[sites.AEMET$siteAQ == st,
                                                                  "indicativo"]),
                                    WorldMet = sites.WorldMet[sites.WorldMet$siteAQ == st,
                                                              "code"]
                                   )
                        )
}

head(nn.stations)

siteAQ  AEMET WorldMet    
1 es0041a 1082  080210-99999
2 es0110a 1082  080210-99999
3 es0118a 3195  082210-99999
4 es0120a 3195  082210-99999
5 es0817a 5783  083830-99999
6 es0890a 5783  083830-99999

## ERA5-Land Data

In [7]:
#-----------------------------
#    ERA5-Land Data
#-----------------------------

folder.ERA5.Land <- "data/Curation/ERA5-Land/"

# Relative Humidity
load(paste(folder.ERA5.Land,
           "rh_daily_2010_2020_final_stations.rda", sep=""))

rh.ERA5.Land <- data.as.datetime(df, "dates", "ymd") %>%
                    add_column(variable="RH", .after="dates")

# Solar radiation
load(paste(folder.ERA5.Land, 
           "ssrd_daily_2010_2020_final_stations.rda", sep=""))

ssrd.ERA5.Land <- data.as.datetime(df, "dates", "ymd") %>%
                    add_column(variable="solar.radiation", .after="dates")

ERA5.Land <- rbind(rh.ERA5.Land, ssrd.ERA5.Land)

## Merge Process

In [8]:
# Folder wherever take downloaded data, if it exists
Mto.files <- "data/Curation/"

data_Mto <- data.frame()

In [9]:
for (st in levels(nn.stations$siteAQ)) {
    
    if (st %in% names(ERA5.Land)) {

        code <- nn.stations[nn.stations$siteAQ == st, "WorldMet"]

        data.WorldMet <- read.csv(paste(Mto.files, "WorldMet/",
                                     code, ".csv", sep=""), stringsAsFactor=F) %>%
                        data.as.datetime("date", "ymd") %>%
                        select(-"code")

        indicativo <- nn.stations[nn.stations$siteAQ == st, "AEMET"]

        data.AEMET <- read.csv(paste(Mto.files, "AEMET/",
                                     indicativo, ".csv", sep=""), stringsAsFactor=F) %>%
                        data.as.datetime("fecha", "ymd") %>%
                        select("fecha", "tmed", "prec", 
                               "tmin", "tmax", "presMax", "presMin"
                               )

        data.row <- merge(x = data.WorldMet, y = data.AEMET,
                            by.x = "date", by.y = "fecha", all = TRUE)

        for (vr in levels(as.factor(ERA5.Land$variable))) {
            data.row <- merge(x = data.row, 
                              y = ERA5.Land[ERA5.Land$variable == vr,
                                            c("dates", st)],
                              by.x = "date", by.y = "dates", all.x = TRUE
                             )
            names(data.row)[ncol(data.row)] <- vr
        }

        data.row[, "site"] <- st
        data_Mto <- rbind(data_Mto, data.row)
    }
}

In [10]:
head(data.row)

date       ws        wd        atmos_pres tmed prec tmin tmax presMax presMin
1 2013-01-01 1.9145833  55.99284 1021.96    13.4 0    10.0 16.9 1017.6  1014.1 
2 2013-01-02 2.5104167  60.44484 1025.94    13.5 0     8.6 18.4 1021.5  1017.3 
3 2013-01-03 3.4173913  85.33514 1032.06    11.8 0     7.4 16.3 1032.5  1021.5 
4 2013-01-04 2.9875000  81.18525 1034.62    11.4 0     6.3 16.6 1033.5  1030.5 
5 2013-01-05 0.5041667 173.50238 1034.18    14.4 0     7.5 21.3 1030.9  1024.5 
6 2013-01-06 0.7250000 172.17114 1030.90    16.0 0    11.6 20.4 1024.5  1020.6 
  RH       solar.radiation site   
1 76.08620  93.79024       es2031a
2 69.56252  60.02178       es2031a
3 75.45028 109.05331       es2031a
4 76.41623  98.05585       es2031a
5 76.10843 112.60032       es2031a
6 74.26767 130.97924       es2031a

## Guardar Datos

```R
save(data_Mto, nn.stations, file="data/meteorology.rda")
```